File: 2015/15.jl
using Combinatorics

struct Ingredient
    name::String
    properties::Vector{Int}
    capacity::Int
    durability::Int
    flavor::Int
    texture::Int
    calories::Int
end

function Ingredient(n, vals...)
    p = [v for v in vals]
    return Ingredient(n, p, vals...)
end

function readInput()
    ings = Vector{Ingredient}()
    f = open("instructions15a.txt", "r")
    #f = open("prova.txt", "r")
    lines = readlines(f)
    for l in lines
        println(l)
        items = split(l, ": ")
        name = items[1]
        subitems = split(items[2], ", ")
        vals = [parse(Int, x[2]) for x in split.(subitems, " ")]
        push!(ings, Ingredient(name, vals...))
    end       
    return ings
end

function recipe(partition, properties)
    scores = properties*partition
    if any(x->x<0, scores)
        return 0
    else 
        return reduce(*, properties*partition)
    end
end

function recipe2(partition, properties)
    scores = properties*partition
    if scores[end] != 500
        return 0
    end
    if any(x->x<0, scores)
        return 0
    else 
        return reduce(*, (properties*partition)[1:end-1])
    end
end

function check(ings::Vector{Ingredient}, N::Int, part=1, n_prop=4)
    props = [ings[i].properties[k] for k in 1:n_prop, i in 1:length(ings)]
    props = reshape(props, (n_prop, length(ings)))
    m = 0
    m_recipe = Nothing
    for p in partitions(N, length(ings))
        for c in permutations(p) 
            if part == 1
                score = recipe(c, props)
            else
                score = recipe2(c, props)
            end
            if score > m
                m = score
                m_recipe = c
            end
        end
    end
    
    println("🎅 $m $m_recipe")
end

ings = readInput()
check(ings, 100)
check(ings, 100, 2, 5)

File: 2015/15.py
import time, numpy as np

from itertools import combinations_with_replacement, permutations, combinations

from utils import readInput

class Ingredient:
    def __init__(self, name, capacity, durability, flavor, texture, calories):
        self.name = name
        self.capacity = capacity
        self.durability = durability
        self.flavor = flavor
        self.texture = texture
        self.calories = calories

    def property(self, spoons):
        return np.array([spoons * self.capacity,
                spoons * self.durability,
                spoons * self.flavor,
                spoons * self.texture])

def loadInput():
    lines = readInput("instructions15a.txt")
    ing = []
    for d in lines:
        d = d.split(" ")
        name = d[0][:-1]
        vals = [int(d[i][:-1]) if d[i].endswith(",") else int(d[i]) for i in (2, 4, 6, 8, 10)] 
        ing.append(Ingredient(name, *vals))
    return ing

def starting_solution(n):
    norm = 100
    solution = np.zeros(shape=(n,))
    for i in range(n-1):
        solution[i] = np.random.randint(norm-solution.sum())
    solution[0] = 27
    solution[-1] = norm - solution.sum()
    return solution.astype('int')

def shift_solution(ingredients, solution, c, max_score):
    if solution[c[0]] > 0 and solution[c[1]] < 100:
        solution[c[0]] -= 1
        solution[c[1]] += 1
        new_score = tot_prop(ingredients, solution)
        if new_score > max_score:
            max_score = new_score
            return (True, max_score)
        else:
            solution[c[0]] += 1
            solution[c[1]] -= 1
    return (False, max_score)
        
def optimize(ingredients):
    max_score = 0
    for _ in range(10):
        solution = starting_solution(len(ingredients))
        score = tot_prop(ingredients, solution)
        max_score = max(score, max_score)
        while True:
            for c in combinations(range(len(ingredients)), 2):
                res = shift_solution(ingredients, solution, c, max_score)
                if res[0]:
                    max_score = res[1]
                    break
                else:
                    c = c[::-1]
                    res = shift_solution(ingredients, solution, c, max_score)
                    if res[0]:
                        max_score = res[1]
                        break
            else:
                break
    return max_score

def tot_prop(ing, q):
    totals = ing[0].property(q[0])
    for i in range(1, len(ing)):
        totals += ing[i].property(q[i])

    totals[totals<0] = 0
    return np.prod(totals)

def part1(ingredients):
    tot = optimize(ingredients)
    print (f"🎄 Part 1: {tot}")

def tot_calories(ing, q):
    return sum([ing[i].calories*q[i] for i in range(len(ing))])

def part2(ing):
    tot = 0
    for c in combinations_with_replacement(range(101), len(ing)):
        if sum(c) != 100:
            continue
        for p in permutations(c, 4):
            if tot_calories(ing, p) != 500:
                continue
            tot = max(tot, tot_prop(ing, p))
    print (f"🎄🎅 Part 2: {tot}")
    
if __name__ == "__main__":
    title = "Day 15: Science for Hungry People"
    sub = "-"*(len(title)+2)

    print()
    print(f" {title} ")
    print(sub)
    
    inputs = loadInput()
    
    t0 = time.time()
    part1(inputs)
    print ("Time: {:.5f}".format(time.time()-t0))
    
    t0 = time.time()
    part2(inputs)
    print ("Time: {:.5f}".format(time.time()-t0))
